packages/overlay/src/OverlayBase.ts

```diff
@@ -0,0 +1,891 @@
+
+export const LONGPRESS_INSTRUCTIONS = {
+    touch: 'Double tap and long press for additional options',
+    keyboard: 'Press Space or Alt+Down Arrow for additional options',
+    mouse: 'Click and hold for additional options',
+};
+
+export class BeforetoggleClosedEvent extends Event {
+    currentState = 'open';
+    newState = 'closed';
+    constructor() {
+        super('beforetoggle', {
+            bubbles: false,
+            composed: false,
+        });
+    }
+}
+
+export class BeforetoggleOpenEvent extends Event {
+    currentState = 'closed';
+    newState = 'open';
+    constructor() {
+        super('beforetoggle', {
+            bubbles: false,
+            composed: false,
+        });
+    }
+}
+
+export const noop = (): void => {
+    return;
+};
+
+/**
+ * Apply a "transitionend" listener to an element that may not transition but
+ * guarantee the callback will be fired either way.
+ *
+ * @param el {HTMLElement} - Target of the "transition" listeners.
+ * @param action {Function} - Method to trigger the "transition".
+ * @param cb {Function} - Callback to trigger when the "transition" has ended.
+ */
+export const guaranteedTransitionend = (
+    el: HTMLElement,
+    action: () => void,
+    cb: () => void
+): void => {
+    const cleanup = (): void => {
+        el.removeEventListener('transitionrun', handleTransitionrun);
+        el.removeEventListener('transitionend', handleTransitionend);
+        cb();
+    };
+    let guarantee2: number;
+    let guarantee3: number;
+    // WebKit fires `transitionrun` a little earlier, so the inner/outer relationship
+    // here allows WebKit to be caught, but doesn't remove the animation listener until
+    // after it would have fired in Chromium.
+    const guarantee1 = requestAnimationFrame(() => {
+        guarantee2 = requestAnimationFrame(() => {
+            guarantee3 = requestAnimationFrame(() => {
+                cleanup();
+            });
+        });
+    });
+    const handleTransitionend = (event: TransitionEvent): void => {
+        if (event.propertyName === 'visibility') {
+            // Ignore "visibility" transitions because they often happen before/after a
+            // larger transition and don't represent the overall transition duration.
+            return;
+        }
+        cleanup();
+    };
+    const handleTransitionrun = (event: TransitionEvent): void => {
+        if (event.propertyName === 'visibility') {
+            // Ignore "visibility" transitions because they often happen before/after a
+            // larger transition and don't represent the overall transition duration.
+            return;
+        }
+        cancelAnimationFrame(guarantee1);
+        cancelAnimationFrame(guarantee2);
+        cancelAnimationFrame(guarantee3);
+        el.removeEventListener('transitionrun', handleTransitionrun);
+        el.addEventListener('transitionend', handleTransitionend);
+    };
+    el.addEventListener('transitionrun', handleTransitionrun);
```

Comment:
Would this work, & also allow us to delete `cleanup` in favor of calling `cb()` directly?

```suggestion
        el.addEventListener('transitionend', handleTransitionend, { once: true });
    };
    el.addEventListener('transitionrun', handleTransitionrun, { once: true });
```

Reply:
Address via:

> The other path I thought of here was to count up the transitionrun events, and then count down the transitionend events, so we could know 100% that ALL of the transitions had completed. Let me know if you think that would be better here than the visibility caveats that currently are leveraged.